src/libsyntax/feature_gate.rs

```diff
@@ -848,13 +863,30 @@ impl<'a> Context<'a> {
             // feature gate checking. Macro gating runs
             // before the plugin attributes are registered
             // so we skip this then
-            if !is_macro {
-                gate_feature!(self, custom_attribute, attr.span,
-                              &format!("The attribute `{}` is currently \
-                                        unknown to the compiler and \
-                                        may have meaning \
-                                        added to it in the future",
-                                       name));
+            if !is_macro && !(self.features.proc_macro || self.features.custom_attribute) {
+                let custom_attrs_issue = find_lang_feature_issue("custom_attribute")
+                                            .expect("Feature `custom_attribute` should have \
+                                                     a tracking issue");
+
+                let proc_macro_issue = find_lang_feature_issue("proc_macro")
+                                        .expect("Feature `proc_macro` should have \
+                                                 a tracking issue");
+
+                let err_msg = format!("The attribute `{}` is currently unknown to the compiler \
+                                        and may have meaning added to it in the future \
+                                        (see issue #{})", name, custom_attrs_issue);
+
+                let mut err = self.parse_sess.span_diagnostic.struct_span_err(attr.span, &err_msg);
+
+                if self.parse_sess.unstable_features.is_nightly_build() {
+                    err.help(&format!("if you meant to use this attribute as a procedural macro \
+                                       (see issue #{}), add `#![feature(proc_macro)]` to the \
+                                       crate attributes to enable procedural macros, \
+                                       or `#![feature(custom_attribute)]` otherwise.",
```

Comment:
I would mention `#![feature(custom_derive)]` first here, something like:
`.help("add #![feature(custom_attribute)] to the crate attributes to enable")` (original message)
``.help("if you meant to use this attribute as a procedural macro, add `#![feature(proc_macro)]` (see issue #{})")``